Fix #943: Incorporate Socket Mode as default for Getting Started guide (#990)
slackapi/bolt-js: docs/_tutorials/getting_started.md

![Basic Information page](../assets/basic-information-page.png "Basic Information page")

Look around, add an app icon and description, and then let's start configuring your app. 🔩

---

### Tokens and installing apps
Slack apps use [OAuth to manage access to Slack's APIs](https://api.slack.com/docs/oauth). When an app is installed, you'll receive a token that the app can use to call API methods. 

> There are two main token types available to a Slack app: user (`xoxp`) and bot (`xoxb`) tokens. 

Reviewer: ```suggestion
There are three main token types available to a Slack app: user (`xoxp`), bot (`xoxb`), and app (`xapp`) tokens. 
```

- User tokens allow you to call API methods on behalf of users after they install or authenticate the app. There may be several user tokens for a single workspace. 

> - Bot tokens are associated with bot users, and are only granted once in a workspace where someone installs the app. The bot token your app uses will be the same no matter which user performed the installation. Bot tokens are the token type that _most_ apps use.

Reviewer: ```suggestion
- Bot tokens are associated with bot users, and are only granted once in a workspace where someone installs the app. The bot token your app uses will be the same no matter which user performed the installation. Bot tokens are the token type that _most_ apps using HTTP connections.
```

![Basic Information page](../assets/basic-information-page.png "Basic Information page")

Look around, add an app icon and description, and then let's start configuring your app. 🔩

---

### Tokens and installing apps
Slack apps use [OAuth to manage access to Slack's APIs](https://api.slack.com/docs/oauth). When an app is installed, you'll receive a token that the app can use to call API methods. 

There are three main token types available to a Slack app: user (`xoxp`), bot (`xoxb`), and app (`xapp`) tokens. 
- [User tokens](https://api.slack.com/authentication/token-types#user) allow you to call API methods on behalf of users after they install or authenticate the app. There may be several user tokens for a single workspace. 
- [Bot tokens](https://api.slack.com/authentication/token-types#bot) are associated with bot users, and are only granted once in a workspace where someone installs the app. The bot token your app uses will be the same no matter which user performed the installation. Bot tokens are the token type that _most_ apps use.

> - [App tokens](https://api.slack.com/authentication/token-types#app) represent your app across organizations, including installations by all individual users on all workspaces in a given organization.

Reviewer 2: We call this "App-level token" in docs.
```suggestion
- [App-level tokens](https://api.slack.com/authentication/token-types#app) represent your app across organizations, including installations by all individual users on all workspaces in a given organization.
```

You’ll be prompted with a series of questions to describe your new project (you can accept the defaults by hitting <kbd>Enter</kbd> on each prompt if you aren’t picky). After you’re done, you’ll have a new `package.json` file in your directory.

> Before we install the Bolt for JavaScript package to your new project, let's save the **bot token** and **Signing Secret** that were generated when you configured your app. These should be stored as local environment variables and should *not* be saved in version control.

Author: I do like the idea of linking out to more resources on app security. I've added this!

> export SLACK_BOT_TOKEN=xoxb-<your-bot-token>

Reviewer: Hmmm...I think it may be a bit confusing to save the bot token here then switch to an app token later

Author: @shaydewael When I remove the `token` field from my Bolt test app and leave only `appToken`, I get an App Initialization error, so my understanding is that you need a bot token and an app token to use Socket Mode. I'm unclear why that is, personally. 

In this case the app token becomes relevant when the user is setting up Socket Mode, so that's why I figured I'd introduce it in the later section. Open to feedback that that's confusing though. We can also guide the user at this stage in getting their app-token. It would also work.

})();
```

Save your `app.js` file, then on the command line run the following:

```script
node app.js
```

Your app should let you know that it's up and running. 🎉 

---

### Setting up events
Your app behaves similarly to people on your team — it can post messages, add emoji reactions, and listen and respond to events. 

To listen for events happening in a Slack workspace (like when a message is posted or when a reaction is posted to a message) you'll use the [Events API to subscribe to event types](https://api.slack.com/events-api). For this guide, we are going to be using [Socket Mode](https://api.slack.com/apis/connections/socket), our recommended option for those just getting started and building something for their team. 

> > 💡 Socket Mode allows your app to use the Events API and interactive components of the platform—without exposing a public HTTP Request URL. This mode is helpful for  development, or if you're receiving requests from behind a firewall. HTTP is more useful for apps being deployed to a hosting environment, or apps intended for distribution via the Slack App Directory. You can easily switch between Socket Mode and a direct HTTP endpoint and at any time in the app config. 

Reviewer: ```suggestion
> 💡 Socket Mode allows apps to use the Events API and interactive components without exposing a public HTTP endpoint. This can be helpful during development, or if you're receiving requests from behind a firewall. HTTP is more useful for apps being deployed to hosting environments (like [AWS](/bolt-js/deployments/aws-lambda) or [Heroku](/bolt-js/deployments/heroku)), or apps intended for distribution. You can easily toggle between Socket Mode and HTTP connections at any time in the app config. 
```

Reviewer: I see you took out the hosting providers line below—while I think it may be helpful to link to the AWS and Heroku guides in this, feel free to leave it out if you think it adds clutter or disagree for another reason.

Okay, let's enable Socket Mode:

1. Head to your app's configuration page (click on the app [from your app management page](https://api.slack.com/apps)). Navigate to **Socket Mode** on the left side menu and toggle to enable. 

2. Go to **Basic Information** and scroll down under the App Token section and click **Generate Token and Scopes** to generate an app token. Add the `connections:write` scope to this token and save the generated `xapp` token, we'll use that in just a moment. 

Finally, it's time to tell Slack what events we'd like to listen for. Under **Event Subscriptions**, toggle the switch labeled **Enable Events**. 

> When an event occurs, Slack will send your app some information about the event, like the user that triggered it and the channel it occurred in. Your app will process the details and can respond accordingly.

Reviewer: ```suggestion
When an event occurs, Slack will send your app information about the event, like the user that triggered it and the channel it occurred in. Your app will process the details and can respond accordingly.
```

Scroll down to **Subscribe to Bot Events**. There are four events related to messages:

> - `message.channels` listens for messages in public channels that your app is added to

Reviewer: ```suggestion
- [`message.channels`](https://api.slack.com/events/message.channels) listens for messages in public channels that your app is added to
```

Reviewer: These links aren't related to your PR, but I just noticed them and felt they may be helpful to link out to.

> - `message.groups` listens for messages in private channels that your app is added to

Reviewer: ```suggestion
- [`message.groups`](https://api.slack.com/events/message.groups) listens for messages in private channels that your app is added to
```

> - `message.im` listens for messages in your app's DMs with users

Reviewer: ```suggestion
- [`message.im`](https://api.slack.com/events/message.im) listens for messages in your app's DMs with users
```

> - `message.mpim` listens for messages in multi-person DMs that your app is added to

Reviewer: ```suggestion
- [`message.mpim`](https://api.slack.com/events/message.mpim) listens for messages in multi-person DMs that your app is added to
```

If you want your bot to listen to messages from everywhere it is added to, choose all four message events. After you’ve selected the events you want your bot to listen to, click the green **Save Changes** button.

Back in your project, make sure to store the `xapp` token you saved earlier in your environment.

```shell
export SLACK_APP_TOKEN=xapp-<your-app-token>
```

> Make a simple change to your basic Bolt initialization code and restart the app.

Reviewer: ```suggestion
Make a simple change to your Bolt initialization code and restart the app.
```

```javascript

> // Initializes your app with your bot token and signing secret

Reviewer: ```suggestion
// Initializes your app with your app token and signing secret
```

})();
```

Save your `app.js` file, then on the command line run the following:

```script
node app.js
```

Your app should let you know that it's up and running. 🎉 

---

### Setting up events
Your app behaves similarly to people on your team — it can post messages, add emoji reactions, and listen and respond to events. 

To listen for events happening in a Slack workspace (like when a message is posted or when a reaction is posted to a message) you'll use the [Events API to subscribe to event types](https://api.slack.com/events-api). For this guide, we are going to be using [Socket Mode](https://api.slack.com/apis/connections/socket), our recommended option for those just getting started and building something for their team. 

> > 💡 Socket Mode allows apps to use the Events API and interactive components without exposing a public HTTP endpoint. This can be helpful during development, or if you're receiving requests from behind a firewall. HTTP is more useful for apps being deployed to hosting environments (like [AWS](/bolt-js/deployments/aws-lambda) or [Heroku](/bolt-js/deployments/heroku)), or apps intended for distribution. You can easily toggle between Socket Mode and HTTP connections at any time in the app config. 

Author: Yep, I've added some specific language to allow the user to "fork" into the alternate guide :)

### Sending and responding to actions

To use features like buttons, select menus, datepickers, modals, and shortcuts, you’ll need to enable interactivity. Head over to **Interactivity & Shortcuts** in your app configuration. 

> > 💡 You'll notice that with Socket Mode on, basic interactivity is enabled for us by default, so no further action here is needed. If you're using HTTP, you'll need to supply a [Request URL](https://api.slack.com/apis/connections/events-api#the-events-api__subscribing-to-event-types__events-api-request-urls) for Slack to send events to.

Reviewer: 🙌

Reviewer 3: If interactivity is auto enabled for socket mode apps (is it?), should we bother keeping lines 204 and 206 here? Side note, 206 can point to the getting started with http guide

When interactivity is enabled, interactions with shortcuts, modals, or interactive components (such as buttons, select menus, and datepickers) will be sent to your app as events. 

Now, let’s go back to your app’s code and add logic to handle those events:

> - First, we'll send a message that contains an interactive component, a button.

Reviewer: ```suggestion
- First, we'll send a message that contains an interactive component (in this case a button).
```

> - Next, listen for the action of a user clicking the button and respond

Reviewer: ```suggestion
- Next, we'll listen for the action of a user clicking the button before responding
```

* Bolt allows you to [call Web API methods](/bolt-js/concepts#web-api) with the client attached to your app. There are [over 220 methods](https://api.slack.com/methods) on our API site.

> * Learn more about the different token types [on our API site](https://api.slack.com/docs/token-types). Your app may need different tokens depending on the actions you want it to perform.

Reviewer: Maybe note that HTTP apps typically use bot tokens here?

* Bolt allows you to [call Web API methods](/bolt-js/concepts#web-api) with the client attached to your app. There are [over 220 methods](https://api.slack.com/methods) on our API site.

> * Learn more about the different token types [on our API site](https://api.slack.com/docs/token-types). Your app may need different tokens depending on the actions you want it to perform. For apps that do not use Socket Mode, typically only a bot (`xoxb`) token is required. 

Reviewer 3: Another good place to link http getting started guide.
